### PR TITLE
Archetype type filter

### PR DESCRIPTION
**Pull request by lily-pytel**
@alexmiddeleer everything seems to be laid out correctly to do the filtering, but the function doesn't actually filter. I've hit a bit of a roadblock. It looks like cards get filtered through graphql as opposed to a function. I've also tried running it in adminer and it doesn't filter properly, so I think the issue is with the changes I added.

This doesn't filter
```
select mythgard.search_decks_nosort(NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,array['MIDRANGE']::mythgard.deckArchetype[], NULL);
```

But this works:
```
select * from mythgard.deck where archetype = array['MIDRANGE']::mythgard.deckArchetype[]
```

It's tempting to make this a text field or filter on the front end, I won't lie. Any ideas?

### CODE REVIEW

**Comment by lily-pytel** [comment]
@alexmiddeleer I made this test function
```
create or replace function mythgard.test_filter(
  archetype mythgard.deckArchetype[])
  returns setof mythgard.deck as $$

    SELECT deck.* FROM mythgard.deck
    WHERE deck.archetype = archetype::mythgard.deckArchetype[]

    GROUP BY deck.id
    LIMIT 2000;
  $$ language sql stable;

select mythgard.test_filter(array['MIDRANGE']::mythgard.deckArchetype[]);
```

to try to isolate it and it just doesn't work... it returns all rows.

**Comment by alexmiddeleer** [comment]
I can take a look

**Comment by alexmiddeleer** [comment]
@lily-pytel I'd suggest making the input a string and trying the `any` operator, perhaps like this.
```
create or replace function mythgard.test_filter(archy Text)
  returns setof mythgard.deck as $$

    SELECT deck.* FROM mythgard.deck
    WHERE archy::mythgard.deckArchetype = any(deck.archetype)

    GROUP BY deck.id
    LIMIT 2000;
  $$ language sql stable;

select mythgard.test_filter('MIDRANGE');
```
See https://www.postgresql.org/docs/current/arrays.html -> Searching in Arrays

**Comment by lily-pytel** [comment]
@alexmiddeleer The problem is that it gets MIDRANGE, but also {CONTROL, MIDRANGE} - still, it's already better than before, when it wasn't working at all. Also how would you search for {CONTROL, MIDRANGE} in this scenario?

**Comment by lily-pytel** [comment]
@alexmiddeleer Weirdly enough, this works:

```
DROP FUNCTION IF EXISTS filter_enum;
DROP TABLE IF EXISTS someTable;
DROP TYPE IF EXISTS enumExample;

CREATE TYPE enumExample AS ENUM ('blue', 'red', 'yellow', 'green');

CREATE TABLE someTable (
  id SERIAL PRIMARY KEY,
  enumColumn enumExample[]
);

INSERT INTO someTable (id, enumColumn) VALUES (1, array['blue']::enumExample[]);
INSERT INTO someTable (id, enumColumn) VALUES (2, array['blue', 'red']::enumExample[]);
INSERT INTO someTable (id, enumColumn) VALUES (3, array['red', 'blue']::enumExample[]);
INSERT INTO someTable (id, enumColumn) VALUES (4, array['blue', 'green']::enumExample[]);
INSERT INTO someTable (id, enumColumn) VALUES (5, array['red', 'yellow']::enumExample[]);

CREATE OR REPLACE function filter_enum(
  enumArray enumExample[])
  returns setof someTable as $$

    SELECT * FROM someTable
    WHERE someTable.enumColumn = enumArray::enumExample[]

  $$ language sql stable;

select filter_enum(array['blue']::enumExample[]);
select filter_enum(array['blue', 'red']::enumExample[]);
select filter_enum(array['green']::enumExample[]);
```

But this still doesn't

```
create or replace function mythgard.test_filter(
  archetype mythgard.deckArchetype[])
  returns setof mythgard.deck as $$

    SELECT * FROM mythgard.deck
    WHERE deck.archetype = archetype::mythgard.deckArchetype[]

  $$ language sql stable;

select mythgard.test_filter(array['MIDRANGE']::mythgard.deckArchetype[]);
select mythgard.test_filter(array['CONTROL', 'MIDRANGE']::mythgard.deckArchetype[]);
```

WHAT IS THE DIFFERENCE???

**Comment by alexmiddeleer** [comment]
I figured we don’t ever need to allow them to search for multiple.

**Comment by lily-pytel** [comment]
@alexmiddeleer Ugh, I found the difference... and it's so stupid I want to cry: https://github.com/mythgard-hub/mythgard-hub/pull/236/commits/74cb562413f3700ddd9f82c140964a652ade3d2b

Gonna add some test for the new filtering and then this feature should be good to go. Only last annoyance is that the deck page doesn't reload properly after an update but that was happening before as well

**Comment by alexmiddeleer** [comment]
Sql is really tricky sometimes.

**Comment by lily-pytel** [comment]
@alexmiddeleer okay, it's ready for review. I'll test everything again once it's merged to the main feature branch. Thanks for looking into it!

File: next/components/deck-search-form-dropdown-filter.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+export default function DeckSearchFormDropdownFilter(props) {
+  const { label, options, onChange, filterValue } = props;
+
+  return (
+    <label className="input-label">
+      <style jsx>{``}</style>
```
Comment (alexmiddeleer) [inline next/components/deck-search-form-dropdown-filter.js]: ✂️ ?

File: next/components/deck-search-form-dropdown-filter.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      {label}
+      <br />
```
Comment (alexmiddeleer) [inline next/components/deck-search-form-dropdown-filter.js]: Is this br needed or would `display: block` be better? I view br tags as a slightly less odious version of inline styles, but they're convenient enough that I wouldn't blacklist them everywhere.

**Comment by alexmiddeleer** [comment]
`Warning: `value` prop on `select` should not be null. Consider using an empty string to clear the component or `undefined` for uncontrolled components. `

Gettings this after navigation from /home to /decks on first attempt

File: next/components/deck-search-form-dropdown-filter.js (hunk 2 above)
Comment (lily-pytel) [reply next/components/deck-search-form-dropdown-filter.js]: It's what we do for other parts of this page, I was mostly trying to be consistent there

File: next/components/deck-search-form-dropdown-filter.js (hunk 1 above)
Comment (lily-pytel) [reply next/components/deck-search-form-dropdown-filter.js]: I was wondering if you would say that. Will remove.

**Comment by lily-pytel** [comment]
> Warning: value prop on select should not be null. Consider using an empty string to clear the component or undefined for uncontrolled components.

I hate this warning a lot. Null is the best value to use there and I wish they would just leave it alone. Will change though.

File: next/components/deck-search-form-dropdown-filter.js (hunk 2 above)
Comment (alexmiddeleer) [reply next/components/deck-search-form-dropdown-filter.js]: Cool, we can leave it in

**Comment by alexmiddeleer** [comment]
This error is the only remaining issue

image

**Comment by lily-pytel** [comment]
@alexmiddeleer I think you're missing the last commit. It looks like the old error message - that line isn't the same after https://github.com/mythgard-hub/mythgard-hub/pull/236/commits/0c0939cce9293c069ea2f7c1102215423bec31a7

**Pull request merged**